### ISSUE TITLE
Allow multiple openauth providers

### ISSUE THREAD

**Issue by ACPK** [lynndylanhurley/devise_token_auth#23]
In my app, I'm using a seperate model for openauth keys so that a user can connect with multiple providers.  Will this be something that you'll be supporting down the road?

**Comment by ACPK** [comment]
@lynndylanhurley  Where their any progress made with this? :)

**Comment by mkhatib** [comment]
fwiw, I've just implemented my own single-account login using any account (email+password, facebook and Google). The single account ID is the email. 

Here's the changes I had to make in order to get this to work, mostly overriding an action in a controller (from devise_token_auth) just to drop some hard-coded things like the `AND provider='email' clause` in session controller to allow non-email provider to login using email+password. And dropping a check for `provider=='email'` in `PasswordsController#update` to allow non-email provider to update and set their passwords.

Here's the change in my project.
https://github.com/manshar/manshar/pull/177/files

**Comment by phs1919** [comment]
I agree with @asoules I am trying to implement multiple providers in reference to https://github.com/intridea/omniauth/wiki/Managing-Multiple-Providers (I think this is plan B @lynndylanhurley ) but I have to override so many controllers and concern (anything which assumes user has provider and uid). I would really appreciate it if the gem is flexible enough to allow multiple providers.

**Comment by harmoney-justins** [comment]
+1 @phs1919 @mkhatib

**Comment by frankjwu** [comment]
@mkhatib One problem I see with using email as the unique account ID is that users might not use the same email address across their different OAuth provider accounts. Some providers also don't provide email (Twitter for example). Your solution appears to work, but to me, it doesn't seem nearly as robust as Plan B suggested by @lynndylanhurley.

@phs1919, were you able to successfully implement that? I'm planning on doing something similar, but I'm a bit wary of overriding so much code.

**Comment by harmoney-justins** [comment]
@frankjwu i've done something similar with https://github.com/intridea/omniauth/wiki/Managing-Multiple-Providers

But i agree plan B is more robust, allow users to associate with a primary account

**Comment by frankjwu** [comment]
@harmoney-justins What do you use for a User's UID and provider? (assuming you now store that information in the Identities table)

**Comment by justinsoong** [comment]
@frankjwu https://gist.github.com/justinsoong/b273626fb8f994dfade3

**Comment by frankjwu** [comment]
@justinsoong I might be missing something, but that looks like standard Devise and not DeviseTokenAuth?

The problem I was referring to earlier is that DeviseTokenAuth uses the `uid` value in the header to lookup and validate tokens (with users identified by a unique `uid` and `provider` pair). So if that information is instead now stored in the Identities table, you'll have to go in and make changes to any code that assumes that the User model has `uid` and `provider`.

**Comment by seddy** [comment]
If anyone comes across this issue, would be good to get feedback on PR #453 to see if that would solve your problems.

**Comment by wangzuo** [comment]
I made some quick changes on https://github.com/wangzuo/devise_token_auth for multiple provider support
